krkn/core/cleanup.py

```diff
@@ -0,0 +1,179 @@
+        
+        sys.excepthook = self._exception_handler
+
+    def register_cleanup(self, func: Callable) -> None:
+        """
+        Register a cleanup callback function.
+        
+        Args:
+            func: A callable that will be invoked during cleanup.
+                  This is a scenario.cleanup() method.
+        """
+        if callable(func):
+            self._cleanup_funcs.append(func)
+            logging.debug(f"Registered cleanup function: {func.__qualname__ if hasattr(func, '__qualname__') else str(func)}")
+
+    def _signal_handler(self, signum: int, frame) -> None:
+        #TODO: frame is not used
+        """
+        Called when a signal (SIGINT/SIGTERM) is received.
+        
+        Args:
+            signum: Signal number
+            frame: Current stack frame
+        """
+        logging.info(f"Received signal {signum}, initiating cleanup...")
+        self._run_cleanup()
+        sys.exit(128 + signum)  # exit code indicating the signal
+
+    def _exception_handler(self, exc_type, exc_value, exc_traceback) -> None:
+        """
+        Called on uncaught exception to ensure cleanup before handling.
+        
+        Args:
+            exc_type: Exception type
+            exc_value: Exception value
+            exc_traceback: Exception traceback
+        """
+        logging.error(f"Unhandled exception ({exc_type.__name__}): {exc_value}")
+        self._run_cleanup()
+        sys.__excepthook__(exc_type, exc_value, exc_traceback)
+
+    def _run_cleanup(self) -> None:
+        """
+        Execute all registered cleanup functions once.
+        
+        This method ensures cleanup runs once even if triggered from multiple sources.
+        After running all cleanup callbacks, it captures diagnostic info.
+        """
+        if self._done:
+            return
+        
+        self._done = True
+        logging.info("Running cleanup callbacks...")
+        
+        for func in list(self._cleanup_funcs):
+            try:
+                func()
+                logging.info(f"Successfully executed cleanup function: {func.__qualname__ if hasattr(func, '__qualname__') else str(func)}")
+            except Exception as e:
+                logging.error(f"Error during cleanup function {func.__qualname__ if hasattr(func, '__qualname__') else str(func)}: {e}")
+        
+        self._capture_diagnostics()
+        
+        logging.info(f"Cleanup completed. Artifacts saved to {self.artifacts_dir}")
+
+    def _capture_diagnostics(self) -> None:
+        """
+        Capture diagnostic info about the cluster.
+        
+        This method runs kubectl commands to gather information about pods, nodes, 
+        and events, and saves them to files in the artifacts directory.
+        """
+        logging.info("Capturing diagnostic information...")
+        
+        try:
+            self._run_and_save_command(["kubectl", "get", "pods", "-A", "-o", "wide"], "all-pods.txt")
```

Comment:
we do not use `kubectl` or any other binaries to collect cluster data, we use kubernetes API instead.